### PR TITLE
feat: use strict typescript

### PR DESCRIPTION
**Pull request by jrandolf-google**
Issues: #6769

### CODE REVIEW

**Review by OrKoN — CHANGES_REQUESTED** [review]
It looks like the installation tests are failing. Let's separate package.json and dependency updates into a separate CL and also split the code changes into multiple CLs for easier review.

File: src/common/AriaQueryHandler.ts [hunk 1]
```diff
@@ -78,17 +88,21 @@ const queryOne = async (
...
   const { name, role } = parseAriaSelector(selector);
   const res = await queryAXTree(exeCtx._client, element, name, role);
-  if (res.length < 1) {
+  if (!res[0] || !res[0].backendDOMNodeId) {
     return null;
   }
   return exeCtx._adoptBackendNodeId(res[0].backendDOMNodeId);
 };
 
+declare global {
```
Comment (OrKoN) [inline src/common/AriaQueryHandler.ts]: Let's not add the types to the global namespace if they are executing/available in the page context?

File: src/common/Browser.ts [hunk 2]
```diff
@@ -572,33 +572,24 @@ export class Browser extends EventEmitter {
...
     } finally {
-      this.removeListener(BrowserEmittedEvents.TargetCreated, check);
-      this.removeListener(BrowserEmittedEvents.TargetChanged, check);
+      this.off(BrowserEmittedEvents.TargetCreated, check);
+      this.off(BrowserEmittedEvents.TargetChanged, check);
     }
 
     async function check(target: Target): Promise<void> {
-      if (await predicate(target)) resolve(target);
+      if (!isResolved && (await predicate(target))) {
```
Comment (OrKoN) [inline src/common/Browser.ts]: this looks like a race condition? multiple promises can enter the branch  while the predicate is being evaluated

File: src/common/DOMWorld.ts [hunk 3]
```diff
@@ -981,6 +981,7 @@ async function waitForPredicatePageFunction(
   if (polling === 'raf') return await pollRaf();
   if (polling === 'mutation') return await pollMutation();
   if (typeof polling === 'number') return await pollInterval(polling);
+  assert(false);
```
Comment (OrKoN) [inline src/common/DOMWorld.ts]: Not needed?

File: src/common/Debug.ts [hunk 4]
```diff
@@ -16,6 +16,11 @@
 
 import { isNode } from '../environment.js';
 
+declare global {
+  // eslint-disable-next-line no-var
+  var __PUPPETEER_DEBUG: string;
```
Comment (OrKoN) [inline src/common/Debug.ts:21]: can it be `const`? (not sure what TS supports here)

Comment (jrandolf) [reply src/common/Debug.ts:21]: It must be `var` because it is in the global scope.

File: src/common/FrameManager.ts [hunk 5]
```diff
@@ -73,7 +73,7 @@ export class FrameManager extends EventEmitter {
   private _frames = new Map<string, Frame>();
   private _contextIdToContext = new Map<string, ExecutionContext>();
   private _isolatedWorlds = new Set<string>();
-  private _mainFrame: Frame;
+  private _mainFrame!: Frame;
```
Comment (OrKoN) [inline src/common/FrameManager.ts]: where is _mainFrame set? it looks like whether we have it is dependent on CDP. Let's keep it optional here and assert it when we need to use it?

File: src/common/JSHandle.ts [hunk 6]
```diff
@@ -402,6 +403,7 @@ export class ElementHandle<
     } = {}
   ): Promise<ElementHandle | null> {
     const frame = this._context.frame();
+    if (!frame) return null;
```
Comment (OrKoN) [inline src/common/JSHandle.ts]: let's assert and error instead as it should not happen?

File: src/common/JSHandle.ts [hunk 7]
```diff
@@ -114,7 +114,7 @@ const applyOffsetsToQuad = (
  *
  * @public
  */
-export class JSHandle<HandleObjectType = unknown> {
+export class JSHandle<HandleObjectType = any> {
```
Comment (OrKoN) [inline src/common/JSHandle.ts]: I think unknown is much safer? Basically any allows us to do anything with the type (including what's not possible). Let's make it work with unknown.

File: src/common/JSHandle.ts [hunk 8]
```diff
@@ -475,6 +477,7 @@ export class ElementHandle<
     } = {}
   ): Promise<ElementHandle | null> {
     const frame = this._context.frame();
+    if (!frame) return null;
```
Comment (OrKoN) [inline src/common/JSHandle.ts]: same here

File: src/common/JSHandle.ts [hunk 9]
```diff
@@ -952,10 +956,10 @@ export class ElementHandle<
 
     const { offsetX, offsetY } = await this._getOOPIFOffsets(this._frame);
     const quad = result.model.border;
-    const x = Math.min(quad[0], quad[2], quad[4], quad[6]);
-    const y = Math.min(quad[1], quad[3], quad[5], quad[7]);
-    const width = Math.max(quad[0], quad[2], quad[4], quad[6]) - x;
-    const height = Math.max(quad[1], quad[3], quad[5], quad[7]) - y;
+    const x = Math.min(quad[0]!, quad[2]!, quad[4]!, quad[6]!);
```
Comment (OrKoN) [inline src/common/JSHandle.ts:949]: can we narrow the type by checking the length instead of doing `!`?

File: src/common/JSHandle.ts [hunk 10]
```diff
@@ -1069,6 +1073,9 @@ export class ElementHandle<
   ): Promise<ElementHandle<T> | null> {
     const { updatedSelector, queryHandler } =
       getQueryHandlerAndSelector(selector);
+    if (!queryHandler.queryOne) {
```
Comment (OrKoN) [inline src/common/JSHandle.ts]: why not assert like for queryAllArray?

File: src/common/JSHandle.ts [hunk 11]
```diff
@@ -1081,6 +1088,9 @@ export class ElementHandle<
   ): Promise<Array<ElementHandle<T>>> {
     const { updatedSelector, queryHandler } =
       getQueryHandlerAndSelector(selector);
+    if (!queryHandler.queryAll) {
```
Comment (OrKoN) [inline src/common/JSHandle.ts]: ditto

File: src/common/JSHandle.ts (hunk 9 above)
Comment (jrandolf) [reply src/common/JSHandle.ts:949]: This is part of `https://www.typescriptlang.org/tsconfig#noUncheckedIndexedAccess` which I've enabled. It helps assert that we know what we are doing when we use `!`.

Checking length does not assert this.

Comment (OrKoN) [reply src/common/JSHandle.ts:949]: I think it does not work quite nice for arrays then. Can we change the quad type to be `[number, number x 4]` then? That should help us avoid all those exclamations? The quad array is fixed size iirc

File: src/common/JSHandle.ts (hunk 10 above)
Comment (jrandolf) [reply src/common/JSHandle.ts]: This is more robust. Do we want to error? What's the expectation?

File: src/common/JSHandle.ts (hunk 11 above)
Comment (jrandolf) [reply src/common/JSHandle.ts]: See above comment.

File: src/common/JSHandle.ts (hunk 10 above)
Comment (OrKoN) [reply src/common/JSHandle.ts]: why is it different from the handling of queryAllArray below?

Comment (jrandolf) [reply src/common/JSHandle.ts]: For that specific instance, there is no canonical choice for a return value, so I assert there.

Comment (OrKoN) [reply src/common/JSHandle.ts]: I think it'd be better to expect an error in this case. If you do $(`aria/smth`) and aria query handler does not support queryOne, I think it's better to error then pretend that the element is not found.

File: src/common/JSHandle.ts (hunk 9 above)
Comment (jrandolf) [reply src/common/JSHandle.ts:949]: It is, but the type itself is generated.

Comment (OrKoN) [reply src/common/JSHandle.ts:949]: hm, I see then.

File: src/common/FrameManager.ts [hunk 12]
```diff
@@ -261,9 +265,10 @@ export class FrameManager extends EventEmitter {
     }
 
     const frame = this._frames.get(event.targetInfo.targetId);
-    const session = Connection.fromSession(this._client).session(
-      event.sessionId
-    );
+    const connection = Connection.fromSession(this._client);
+    if (!connection) return;
```
Comment (OrKoN) [inline src/common/FrameManager.ts]: should we throw an error here instead? It looks like we normally expect the connection and session below to be there (unless smth goes very wrong).

**Review by OrKoN — APPROVED** [review]

**Pull request merged**